pkg/controller/statefulset/stateful_pod_control.go

```diff
@@ -27,59 +27,101 @@ import (
+type StatefulPodControl struct {
+	objectMgr StatefulPodControlObjectManager
+	recorder  record.EventRecorder
+}
+
+// NewStatefulPodControl constructs a StatefulPodControl using a realStatefulPodControlObjectManager with the given
+// clientset, listers and EventRecorder.
+func NewStatefulPodControl(
 	client clientset.Interface,
-	setLister appslisters.StatefulSetLister,
 	podLister corelisters.PodLister,
-	pvcLister corelisters.PersistentVolumeClaimLister,
+	claimLister corelisters.PersistentVolumeClaimLister,
 	recorder record.EventRecorder,
-) StatefulPodControlInterface {
-	return &realStatefulPodControl{client, setLister, podLister, pvcLister, recorder}
+) *StatefulPodControl {
+	return &StatefulPodControl{&realStatefulPodControlObjectManager{client, podLister, claimLister}, recorder}
 }
 
-// realStatefulPodControl implements StatefulPodControlInterface using a clientset.Interface to communicate with the
-// API server. The struct is package private as the internal details are irrelevant to importing packages.
-type realStatefulPodControl struct {
-	client    clientset.Interface
-	setLister appslisters.StatefulSetLister
-	podLister corelisters.PodLister
-	pvcLister corelisters.PersistentVolumeClaimLister
-	recorder  record.EventRecorder
+// NewStatefulPodControlFromManager creates a StatefulPodControl using the given StatefulPodControlObjectManager and recorder.
+func NewStatefulPodControlFromManager(om StatefulPodControlObjectManager, recorder record.EventRecorder) *StatefulPodControl {
+	return &StatefulPodControl{om, recorder}
+}
+
+// realStatefulPodControlObjectManager uses a clientset.Interface and listers.
+type realStatefulPodControlObjectManager struct {
+	client      clientset.Interface
+	podLister   corelisters.PodLister
+	claimLister corelisters.PersistentVolumeClaimLister
+}
+
+func (om *realStatefulPodControlObjectManager) CreatePod(pod *v1.Pod) error {
+	_, err := om.client.CoreV1().Pods(pod.Namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
+	return err
+}
+
+func (om *realStatefulPodControlObjectManager) GetPod(namespace, podName string) (*v1.Pod, error) {
+	return om.podLister.Pods(namespace).Get(podName)
+}
+
+func (om *realStatefulPodControlObjectManager) UpdatePod(pod *v1.Pod) error {
+	_, err := om.client.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
+	return err
+}
+
+func (om *realStatefulPodControlObjectManager) DeletePod(pod *v1.Pod) error {
+	return om.client.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})
+}
+
+func (om *realStatefulPodControlObjectManager) CreateClaim(claim *v1.PersistentVolumeClaim) error {
+	_, err := om.client.CoreV1().PersistentVolumeClaims(claim.Namespace).Create(context.TODO(), claim, metav1.CreateOptions{})
+	return err
+}
+
+func (om *realStatefulPodControlObjectManager) GetClaim(namespace, claimName string) (*v1.PersistentVolumeClaim, error) {
+	if namespace == "" {
```

Comment:
How did this work previously?

Reply:
I have no idea. This was also something that was necessary when running in a real cluster.

Maybe in a real cluster a default namespace is always set explicitly, so that an empty namespace never happens?

Basically I'm fogged though.

Reply:
Is it possible that this never worked--ie the get call for the claim never worked in the default namespace, and the PVC was just created, but the create is idempotent? The only place it was used is in createPersistentVolumeClaims().

If one tries to create an object twice, is it an error, or will it succeed if the object already exists? IIUC, Create is just a PUT, so maybe that's what's going on.

Reply:
In the unittests, objects are created with no namespace specified, and I guess since the apiserver is mocked out that doesn't get replaced with default.

Reply:
We generally try to avoid doing this defaulting in "real code" - is there a reason the test can't specify namespace?

Reply:
I'll test this in a cluster to see if it's a problem in real life. I wonder if it's related to the confusion around how I'm using the object cache that you've pointed out below.

Reply:
¯\_(ツ)_/¯ this no longer seems to be necessary, either in the e2e tests or the unit tests. Thanks for pointing it out.